### PR TITLE
Set `ModelMetadata.TemplateHint` based on data annotations

### PR DESCRIPTION
**Pull request by dougbu**
- add `CachedDataAnnotationsMetadataAttributes.UIHint`
- set `ModelMetadata.TemplateHint` using `UIHintAttribute` or `HiddenInputAttribute`
- add doc comments for `TemplateHint`-related properties and methods
- add unit tests and use these attributes in functional tests

nits:
- cache and seal `CachedModelMetadata.IsCollectionType`
- correct doc comments for `ModelMetadata.RealModelType`
- add doc comments for `IsCollectionType`-related properties and methods
- add doc comments for `IsComplexType`-related properties and methods
- move `CachedModelMetadata.IsComplexType` right below `IsCollectionType`
 - same for related fields and methods

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/ModelMetadata.cs [hunk 1]
```diff
@@ -238,8 +252,8 @@ public string PropertyName
         protected IModelMetadataProvider Provider { get; set; }
 
         /// <returns>
-        /// Gets <c>T</c> if <see cref="ModelType"/> is <see cref="Nullable{T}"/>;
-        /// <see cref="ModelType"/> otherwise.
+        /// Gets runtime <see cref="Type"/> of <see cref="Model"/> if <see cref="Model"/> is non-<c>null</c> and
+        /// <see cref="ModelType"/> is not <see cref="Nullable{T}"/>; <see cref="ModelType"/> otherwise.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/ModelMetadata.cs]: @rynowak this was one of the more misleading doc comments we discussed recently...

File: test/WebSites/MvcTagHelpersWebSite/Models/Person.cs [hunk 2]
```diff
@@ -28,6 +30,7 @@ public string Password
         }
 
         [EnumDataType(typeof(Gender))]
+        [UIHint("GenderUsingTagHelpers")]
```
Comment (dougbu) [inline test/WebSites/MvcTagHelpersWebSite/Models/Person.cs]: already have a .cshtml file containing `@Html.EditorFor(m => m.Gender)`.  this hint works with the renamed template file to ensure that template is still found.

File: test/WebSites/MvcTagHelpersWebSite/Models/Person.cs [hunk 3]
```diff
@@ -2,11 +2,13 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System.ComponentModel.DataAnnotations;
+using Microsoft.AspNet.Mvc;
 
 namespace MvcTagHelpersWebSite.Models
 {
     public class Person
     {
+        [HiddenInput(DisplayValue = false)]
```
Comment (dougbu) [inline test/WebSites/MvcTagHelpersWebSite/Models/Person.cs]: note updates here result in the same generated HTML.  in this case, the new `@Html.EditorFor(m => m.Employee.Number)` still generates a hidden input.

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/CachedDataAnnotationsModelMetadata.cs [hunk 4]
```diff
@@ -333,6 +333,31 @@ protected override bool ComputeShowForEdit()
...
+        protected override string ComputeTemplateHint()
+        {
+            if (PrototypeCache.UIHint != null)
+            {
+                return PrototypeCache.UIHint.UIHint;
+            }
+
+            if (PrototypeCache.HiddenInput != null)
+            {
+                return "HiddenInput";
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/CachedDataAnnotationsModelMetadata.cs]: `nameof(PrototypeCache.HiddenInput)`?

**Comment by pranavkm** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/CachedDataAnnotationsModelMetadata.cs (hunk 4 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/CachedDataAnnotationsModelMetadata.cs]: :-1: the hint name is only coincidentally the same as the property in the cache.

**Comment by dougbu** [comment]
commit 0549769fd0e467d6f6c96ef59b32c7700461336f

**Comment by dougbu** [comment]
fixes #963

**Pull request merged**